core/trino-main/src/main/java/io/trino/connector/CatalogPruneTask.java

```diff
@@ -0,0 +1,194 @@
+import org.weakref.jmx.Nested;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static com.google.common.net.MediaType.JSON_UTF_8;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static io.airlift.http.client.JsonBodyGenerator.jsonBodyGenerator;
+import static io.airlift.http.client.Request.Builder.preparePost;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class CatalogPruneTask
+{
+    private static final Logger log = Logger.get(CatalogPruneTask.class);
+    private static final JsonCodec<List<CatalogHandle>> CATALOG_HANDLES_CODEC = JsonCodec.listJsonCodec(CatalogHandle.class);
+
+    private final TransactionManager transactionManager;
+    private final CoordinatorDynamicCatalogManager catalogManager;
+    private final NodeInfo nodeInfo;
+    private final ServiceSelector selector;
+    private final HttpClient httpClient;
+
+    private final ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1, daemonThreadsNamed("catalog-prune"));
+    private final ThreadPoolExecutorMBean executorMBean = new ThreadPoolExecutorMBean(executor);
+
+    private final boolean isEnabled;
+    private final Duration updateInterval;
+    private final boolean httpsRequired;
+
+    private final AtomicBoolean started = new AtomicBoolean();
+
+    @Inject
+    public CatalogPruneTask(
+            TransactionManager transactionManager,
+            CoordinatorDynamicCatalogManager catalogManager,
+            NodeInfo nodeInfo,
+            @ServiceType("trino") ServiceSelector selector,
+            @ForNodeManager HttpClient httpClient,
+            CatalogPruneTaskConfig catalogPruneTaskConfig,
+            InternalCommunicationConfig internalCommunicationConfig)
+    {
+        this.transactionManager = transactionManager;
+        this.catalogManager = requireNonNull(catalogManager, "catalogManager is null");
+        this.nodeInfo = requireNonNull(nodeInfo, "nodeInfo is null");
+        this.selector = requireNonNull(selector, "selector is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+
+        this.isEnabled = catalogPruneTaskConfig.isEnabled();
+        updateInterval = catalogPruneTaskConfig.getUpdateInterval();
+
+        this.httpsRequired = internalCommunicationConfig.isHttpsRequired();
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (isEnabled && started.compareAndSet(false, true)) {
```

Comment:
Nit: using `getAndSet()` seems cleaner here, since the code will read as "if not started"
```java
if (enabled && !started.getAndSet(true)) {